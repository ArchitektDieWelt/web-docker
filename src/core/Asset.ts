export type AssetPosition = "head" | "body";

export type AssetBase = {
  src: string;
  name?: string;
  position?: AssetPosition;
  priority?: number;
  countries?: string[];
};

export type AssetScript = AssetBase & {
  type: "js";
  async?: boolean;
  buildType?: "modern" | "legacy";
  deferred?: boolean;
};

export type AssetLink = AssetBase & {
  type: "css";
  media?: string;
};

export type Asset = AssetLink | AssetScript;
