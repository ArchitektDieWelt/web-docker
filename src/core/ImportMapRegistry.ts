interface ImportMap {
  [key: string]: string;
}

export class ImportMapRegistry {
  private importMap: ImportMap = {};

  constructor(
    private readonly script: HTMLScriptElement = document.createElement("script"),
    private readonly documentHead = document.head
  ) {
    this.script.type = "importmap";
    this.documentHead.appendChild(this.script);
  }

  public add(importMap: ImportMap) {
    this.importMap = {
      ...this.importMap,
      ...importMap,
    };
    this.script.textContent = JSON.stringify(this.importMap);
  }

  public get(): ImportMap {
    return this.importMap;
  }
}
