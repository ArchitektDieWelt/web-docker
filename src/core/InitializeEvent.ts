
export const InitializeEventType = "web-docker:initialize" as const;

class InitializeEvent extends CustomEvent<{
  configFilePath: string;
  logEvents: boolean;
}> {
  constructor(detail: { configFilePath: string; logEvents: boolean }) {
    super(InitializeEventType, { detail });
  }
}

export default InitializeEvent;
