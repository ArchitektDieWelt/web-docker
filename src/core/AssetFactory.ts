import { Asset, AssetLink, AssetPosition, AssetScript } from "~/core/Asset";

export default class AssetFactory {
  private static buildTagScript(attr: AssetScript): HTMLScriptElement {
    const script = document.createElement("script");
    script.src = attr.src;
    if (attr.buildType === "modern") {
      script.type = "module";
      script.setAttribute("crossorigin", "");
    } else {
      script.type = "text/javascript";
      if (attr.async) script.setAttribute("async", "");
    }

    if (attr.priority)
      script.setAttribute("data-priority", attr.priority.toString());

    return script;
  }

  private static buildTagLink(attr: AssetLink): HTMLLinkElement {
    const link = document.createElement("link");
    link.setAttribute("rel", "stylesheet");
    link.setAttribute("href", attr.src);
    if (attr.media) link.setAttribute("media", attr.media);

    if (attr.priority)
      link.setAttribute("data-priority", attr.priority.toString());

    return link;
  }

  public create(
    assets: Asset[],
    position: AssetPosition = "head",
    countryCode?: string
  ): (HTMLLinkElement | HTMLScriptElement)[] {
    return (
      countryCode
        ? assets.filter(
            (asset) =>
              typeof asset.countries !== "object" ||
              asset.countries.includes(countryCode)
          )
        : assets
    )
      .filter((asset) => (asset.position ?? "head") === position)
      .sort((a, b) => {
        if (!a.priority || !b.priority) return 0;
        if (a.priority < b.priority) {
          return 1;
        } else if (a.priority > b.priority) {
          return -1;
        }
        return 0;
      })
      .map((entry) => {
        switch (entry.type) {
          case "css":
            return AssetFactory.buildTagLink(entry);
          case "js":
            return AssetFactory.buildTagScript(entry);
        }
      })
      .filter((item): item is HTMLLinkElement | HTMLScriptElement => !!item);
  }
}
