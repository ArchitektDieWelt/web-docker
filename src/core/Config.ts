import { ModuleConfig } from "~/core/ModuleConfig";

export type Config = ModuleConfig;
