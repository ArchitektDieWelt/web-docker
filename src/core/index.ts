import RegisterEvent, { RegisterEventType } from "~/core/RegisterEvent";
import InitializeEvent, { InitializeEventType } from "~/core/InitializeEvent";

export { default as initialize } from "./initialize";

declare global {
  interface WindowEventMap {
    [RegisterEventType]: RegisterEvent;
    [InitializeEventType]: InitializeEvent;
  }
  interface Window {
    webdocker: {
      initialize: (config: {
        configFilePath: string;
      }) => void;
    };
  }
}
