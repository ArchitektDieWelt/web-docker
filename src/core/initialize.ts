import {
  RemoteConfig,
  RemoteConfigurationService,
} from "~/core/RemoteConfigurationService";
import ModuleRegistry from "~/core/ModuleRegistry";
import { Config } from "~/core/Config";
import { ModuleConfigService } from "~/core/ModuleConfig";
import { RegisterEventType } from "~/core/RegisterEvent";
import { InitializeEventType } from "~/core/InitializeEvent";

const initialize = async (remoteConfig: RemoteConfig) => {
  const moduleConfigService = new ModuleConfigService();
  const registry = new ModuleRegistry();
  const remoteConfigurationService = new RemoteConfigurationService(
    remoteConfig
  );
  // NOTE: event registry has to be done before any async operation. Since Async operations
  // do not block the event loop and the registration event might be dispatched while another event loop is running.
  window.addEventListener(
    RegisterEventType,
    (event: CustomEvent<Config>) => {
      const config = moduleConfigService.getModuleConfig(event.detail);
      registry.addReplace(config);
    }
  );

  const remoteConfigurations = await remoteConfigurationService.fetch();
  remoteConfigurations?.forEach((config: Config) => {
    const moduleConfig = moduleConfigService.getModuleConfig(config);
    registry.add(moduleConfig);
  });
};

window.openreply = { initialize };

window.addEventListener(
  InitializeEventType,
  async (event: CustomEvent<RemoteConfig>) => {
    await initialize(event.detail);
  }
);

export default initialize;
