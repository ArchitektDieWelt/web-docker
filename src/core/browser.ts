import { InitializeEventType } from "~/core/InitializeEvent";
import initialize, { WebDockerOptions } from "~/core/initialize";

window.webdocker = { initialize };

window.addEventListener(
  InitializeEventType,
  async (event: CustomEvent<WebDockerOptions>) => {
    await initialize(event.detail);
  }
);

