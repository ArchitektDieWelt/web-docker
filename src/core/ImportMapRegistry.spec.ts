import { beforeEach, describe, expect, it } from "vitest";
import { ImportMapRegistry } from "~/core/ImportMapRegistry";

describe("ImportMapRegistry", () => {
  beforeEach(() => {
    document.body.innerHTML = "";
    document.head.innerHTML = "";
  });
  it("constructs", () => {
    const service = new ImportMapRegistry();

    expect(service).toBeTruthy();
  });

  it("injects import map to page", async () => {
    window.location.href = "http://localhost/match";

    const service = new ImportMapRegistry();

    expect(document.head).toMatchSnapshot();
  });

  it("injects new entry to import map", async () => {
    window.location.href = "http://localhost/match";

    const service = new ImportMapRegistry();

    service.add({
      "test-module": "/test-module.js",
    });

    expect(document.head).toMatchSnapshot();
  });

  it("injects multiple entries to import map", async () => {
    window.location.href = "http://localhost/match";

    const service = new ImportMapRegistry();

    service.add({
      "test-module": "/test-module.js",
      "test-module2": "/test-module2.js",
    });

    expect(document.head).toMatchSnapshot();
  });
});
