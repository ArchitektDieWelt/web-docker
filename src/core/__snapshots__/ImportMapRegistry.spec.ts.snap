// Vitest Snapshot v1

exports[`ImportMapRegistry > injects import map to page 1`] = `
<head>
  <script
    type="importmap"
  />
</head>
`;

exports[`ImportMapRegistry > injects multiple entries to import map 1`] = `
<head>
  <script
    type="importmap"
  >
    {"test-module":"/test-module.js","test-module2":"/test-module2.js"}
  </script>
</head>
`;

exports[`ImportMapRegistry > injects new entry to import map 1`] = `
<head>
  <script
    type="importmap"
  >
    {"test-module":"/test-module.js"}
  </script>
</head>
`;
